### PR TITLE
[RISCV][WIP] Let RA do the CSR saves.

### PR DESCRIPTION
**Pull request by mgudim**
We turn the problem of saving and restoring callee-saved registers efficiently into a register allocation problem. This has the advantage that the register allocator can essentialy do shrink-wrapping on per register basis. Currently, shrink-wrapping pass saves all CSR in the same place which may be suboptimal. Also, improvements to register allocation / coalescing will translate to improvements in shrink-wrapping.

In `finalizeLowering()` we copy all callee-saved registers from a physical register to a virtual one. In all return blocks we copy do the reverse.

### CODE REVIEW

**Comment by mgudim** [comment]
Currently the shrink wrapping saves / restores all registers at the same point. We noticed that this is not optimal in several places in SPEC. GCC can do this on per-register basis and there is an option to turn this off. By using this flag in gcc we determined that this can make up to 2% difference in dynamic instruction count on gcc benchmark and some smaller impacts in other benchmarks.

Instead of trying to improve shrink-wrapping in llvm, I decided to try this (crazy) idea since it was pretty easy to implement this prototype. I disabled this for functions which can unwind, because this is just an experiment to see if this is viable. Thus all c++ benchmarks that have exceptions were not affected. I got 1.7% improvement on gcc and 1.7% degradation on xz in dynamic instruciton count.

Do you think it's worth pursuing this idea further? To fix the degradations, we need to improve register allocator / coalescing which may have more impact than just improving shrink wrapping.

**Comment by topperc** [comment]
Will this still work with -msave-restore and cm.push/pop?

**Comment by topperc** [comment]
I fought some old post about improving shrink wrapping. https://github.com/sifive/riscv-llvm-internal/pull/1932 not sure how much of it was implemented.

CC @francisvm

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -21314,6 +21314,70 @@ unsigned RISCVTargetLowering::getCustomCtpopCost(EVT VT,
...
+  BitVector MustCalleeSavedRegs;
+  TFI.determineMustCalleeSaves(MF, MustCalleeSavedRegs);
+  const MCPhysReg *CSRegs = MF.getRegInfo().getCalleeSavedRegs();
+  SmallVector<MCPhysReg, 4> EligibleRegs;
+  for (int i = 0; CSRegs[i]; ++i) {
+    if (!MustCalleeSavedRegs[i])
+      EligibleRegs.push_back(CSRegs[i]);
+  }
+
+  dbgs() << "EligibleRegs: " << EligibleRegs.size() << "\n";
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21346]: please wrap this with `LLVM_DEBUG`.

**Comment by mshockwave** [comment]
Does this update any existing tests?

**Comment by topperc** [comment]
> Does this update any existing tests?

It's gated by a command line flag that isn't enabled by default.

**Comment by francisvm** [comment]
> I fought some old post about improving shrink wrapping. https://lists.llvm.org/pipermail/llvm-dev/2017-August/116137.html not sure how much of it was implemented.
> 
> CC @francisvm

Not much of it was implemented in the end. It turned out that for us, the compile-time impact, the unwinding impact and the various regressions were not worth the speedup in the few benchmarks. I wish I had the old data to justify this, but that was a while ago. The unwinder impact was more significant than I initially thought at that time, since Apple platforms rely heavily on compact unwinding, therefore missing some frames and falling back to eh/debug_frame CFI is a huge penalty, and in some memory/permission-constrained environments, very very hard, or impossible for those that only look at prologues/epilogues to quickly unwind (some crash reporters, for example). That made enabling it by default a hard task, and we never pursued an attribute/flag-based approach.

**Comment by mgudim** [comment]
> Will this still work with -msave-restore and cm.push/pop?

I think so. All it does is it makes a callee-saved register a spill, so this should work. I'll run spec with these options to check. If we need to optimize for size, than we can just turn this off with a flag.

**Comment by mgudim** [comment]
At this point I just want to understand if I should continue with this approach or not. If "yes", I'll investigate the degradation on xz, figure out what goes wrong if function unwinds and add tests.

**Comment by topperc** [comment]
> At this point I just want to understand if I should continue with this approach or not. If "yes", I'll investigate the degradation on xz, figure out what goes wrong if function unwinds and add tests.

I think you should continue.

I'm also curious how this interacts with -fno-omit-frame-pointer which makes the X8 register reserved, but its still in the callee save list.

**Comment by wangpc-pp** [comment]
I support to continue to investigate where we can reach via this approach, I think this can be a good workaroud to enhance "shrink wrapping". Thanks for your great idea and I will do some evaluations too.

**Comment by mgudim** [comment]
Hi all, just just wanted to give an update on this:

I looked at the xz degradation:

Half of it is due to the fact that since now we have some long live ranges, early-machinelicm doesn't hoist as well as before. This is easy to fix though. If the register pressure for some class exceeds the limit then there will be a spill. If there are live ranges that are live-through entire loop without any uses inside the loop, then these ranges will be spilled first. So basically, I fixed this by ignoring such live-through-no-uses-inside-loop live ranges in `InitRegPressure`.

The other problem is that RA seems to be doing something wrong and I got some extra spills. I hope that can be fixed too.

The biggest problem is that we need to emit CFI directives that tell runtime how to restore ALL callee-saved registers. With my approach I don't know what is the best way to do this. I don't know this CFI stuff very well, so I am hoping someone can give an advise / suggestion. I can only think of having a pass at the very end of the pipeline which tracks what happens to callee-saved registers and emits the directives accordingly. BTW, there is something which looks relevant in `CFIFixup` and `CFIInstrInserter`, but I haven't looked closely yet.

@topperc @preames @mshockwave @wangpc-pp @asb any suggestions how to do this?

**Comment by wangpc-pp** [comment]
How will GCC handle these CFIs?

**Comment by qcolombet** [comment]
Hi @mgudim 

> At this point I just want to understand if I should continue with this approach or not. If "yes", I'll investigate the degradation on xz, figure out what goes wrong if function unwinds and add tests.

The approach is not crazy at all!
In fact, I believe LLVM current approach is the crazy one :).
I believe you will have a bunch of regressions as it stands though. Up to you to pursue.
Since you're doing that only for 1 target and not all of them, you may have an easier time to track and fix the regression (as opposed to changing RA for all targets at once.)

FWIW, I believe the whole CSR distinction in LLVM is wrong and the proper way to handle them would be to issue parallel copies in the entry block and exit blocks right during ISel (at the very beginning).
In a sense what you are doing is reproducing exactly that, although you do not use parallel copies.

As you saw, this change in representation creates regression because the CSR distinction is baked left and right in the backend. So we can fix them as we find them, but there'll be a long tail.

One remark, since you're not using parallel copies, you're creating artificial interferences and that can have a cascading effect. (See below.) Even if you were to use parallel copies (in LLVM that would be bundles of copies) that wouldn't work out-of-the-box either since this constructs is not a first class citizen.

The TL;DR is you're testing the RA heuristic in ways that has not been exercised and some things will get better, some things will get worse.

Example of issue with CSR using sequential copies
```
v1 = COPY csr1
v2 = COPY csr2
v3 = COPY csr3
```

If for some reason `csr3` gets used for something else, you are at the mercy of the RA ordering for the coalescing. I.e., worst case, you may end up with:
```
some_reg = COPY csr1
csr1 = COPY csr2
csr2 = COPY csr3
```
instead of
```
csr1 = COPY csr1 # no-op
csr2 = COPY csr2 # no-op
some_reg = COPY csr3
```

**Comment by mgudim** [comment]
@qcolombet 

Hi Quentin,

Thank you for the comments. After I looked into some degradations, I see that you are exactly right!

> proper way to handle them would be to issue parallel copies

Interesting, you mean we should put all the copies into one instruction bundle?

> As you saw, this change in representation creates regression because the CSR distinction is baked left and right in the backend. So we can fix them as we find them, but there'll be a long tail.

In particular in the register allocator I see that CSRs are treated differently: since there is some cost to use CSR for the first time, there is some logic trying to minimize these costs. In this approach, I think this logic is working against us. 

Here's another problem that I saw: when trying to allocate for the live range of `x1` (the return-address register on RISCV), the allocator thinks its a good idea to evict previously allocated CSR's live range (even if it means breaking a hint). This eventually leads to  many unnecessary spills. Looks like this is exactly the situation you described here:

> Example of issue with CSR using sequential copies

v1 = COPY csr1
v2 = COPY csr2
v3 = COPY csr3
If for some reason csr3 gets used for something else, you are at the mercy of the RA ordering for the coalescing. I.e., worst case, you may end up with:

some_reg = COPY csr1
csr1 = COPY csr2
csr2 = COPY csr3
instead of

csr1 = COPY csr1 # no-op
csr2 = COPY csr2 # no-op
some_reg = COPY csr3

I have tried some hacks which basically disable eviction if it breaks hints and they make the problem go away, but I am not sure what the proper solution to this should be, maybe we can discuss it during your office hour.

**Comment by qcolombet** [comment]
> @qcolombet
> 
> Hi Quentin,
> 
> Thank you for the comments. After I looked into some degradations, I see that you are exactly right!
> 
> > proper way to handle them would be to issue parallel copies
> 
> Interesting, you mean we should put all the copies into one instruction bundle?

Yes, but that will not be enough because RA needs to be taught how to break the bundle if some of the copies need to be spilled.

> 
> > As you saw, this change in representation creates regression because the CSR distinction is baked left and right in the backend. So we can fix them as we find them, but there'll be a long tail.
> 
> In particular in the register allocator I see that CSRs are treated differently: since there is some cost to use CSR for the first time, there is some logic trying to minimize these costs. In this approach, I think this logic is working against us.

Not surprised and that's what I was referring to when I was saying that CSRs are treated differently left and right.

> 
> Here's another problem that I saw: when trying to allocate for the live range of `x1` (the return-address register on RISCV), the allocator thinks its a good idea to evict previously allocated CSR's live range (even if it means breaking a hint). This eventually leads to many unnecessary spills. Looks like this is exactly the situation you described here:
> 
> > Example of issue with CSR using sequential copies
> 
> v1 = COPY csr1 v2 = COPY csr2 v3 = COPY csr3 If for some reason csr3 gets used for something else, you are at the mercy of the RA ordering for the coalescing. I.e., worst case, you may end up with:
> 
> some_reg = COPY csr1 csr1 = COPY csr2 csr2 = COPY csr3 instead of
> 
> csr1 = COPY csr1 # no-op csr2 = COPY csr2 # no-op some_reg = COPY csr3
> 
> I have tried some hacks which basically disable eviction if it breaks hints and they make the problem go away, but I am not sure what the proper solution to this should be, maybe we can discuss it during your office hour.

Sadly I don't think there's a good solution here. RA's decisions are currently too narrow-scoped to reason about this kind of things. The eviction chain heuristic is supposed to help with these cases, but this is a heuristic! (Plus it may be disabled by default, I don't remember.)

Long story short, if we want to go down that path, I think that a more serious overhaul of the RA infrastructure is needed.

**Comment by michaelmaitland** [comment]
I think something went wrong on the last push.

**Comment by mgudim** [comment]
> I think something went wrong on the last push.

Yes, thanks! I am working on cleaning up the code and posting the branch so others can try it. Then I'll break it up into commits and try to merge the one-by-one

**Comment by mgudim** [comment]
> I think something went wrong on the last push.

Yes, thanks! I am working on cleaning up the code and posting the branch so others can try it. Then I'll break it up into commits and try to merge the one-by-one

**Comment by mgudim** [comment]
I've updated the branch with the latest prototype. To try the branch, add the following flags to clang:
`-mllvm -riscv-enable-save-csr-in-ra=true `  
`mllvm -min-weight-ratio-needed-to-evict-hint=7.5` - this controls a new heuristic I added to register allocator. I've arrived at `7.5` by experiment, so other value could work better in your case.

Looking forward to your feedback. If there are no principal issues moving forward I can start upstreaming this work in small commits.

**Comment by michaelmaitland** [comment]
What is the thought process of taking this approach over modifying the ShrinkWrap pass to communicate the PrologEpilogInserter where to put each callee saved register?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -21874,6 +21874,108 @@ bool RISCVTargetLowering::fallBackToDAGISel(const Instruction &Inst) const {
   return false;
 }
 
+static MachineInstr *findInstrWhichNeedAllCSRs(MachineBasicBlock &MBB) {
+  // Some instructions may require (implicitly) all CSRs to be saved.
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21878]: Is it possible that there are other functions in addition to these functions that need all CSRs to be saved? Is it possible that there can be multiple MI in MBB that need this kind of handling?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -21874,6 +21874,108 @@ bool RISCVTargetLowering::fallBackToDAGISel(const Instruction &Inst) const {
...
+    if (MBB.isReturnBlock())
+      RestorePoints.push_back(&MBB.back());
+    if (MachineInstr *CallToCxaThrow = findInstrWhichNeedAllCSRs(MBB)) {
+      // MachineBasicBlock::iterator MII = CallToCxaThrow->getIterator();
+      //++MII;
+      // assert(MII->getOpcode() == RISCV::ADJCALLSTACKUP && "Unexpected
+      // instruction");
+      //++MII;
+      MachineBasicBlock::iterator MII = MBB.getFirstTerminator();
+      MachineInstr *NewRetMI = BuildMI(MBB, MII, CallToCxaThrow->getDebugLoc(),
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21926]: The introduction of `UnreachableRET` is interesting to me. What is the purpose of it? Why is it relevant for this work? Is it useful in any other instances where we have a return after a call to a function marked `noreturn`?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 4]
```diff
@@ -21874,6 +21874,108 @@ bool RISCVTargetLowering::fallBackToDAGISel(const Instruction &Inst) const {
...
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  const TargetInstrInfo &TII = *MF.getSubtarget().getInstrInfo();
+  const RISCVRegisterInfo &TRI = *Subtarget.getRegisterInfo();
+  const RISCVFrameLowering &TFI = *Subtarget.getFrameLowering();
+
+  SmallVector<MachineInstr *, 4> RestorePoints;
+  SmallVector<MachineBasicBlock *, 4> SaveMBBs;
+  SaveMBBs.push_back(&MF.front());
+  for (MachineBasicBlock &MBB : MF) {
+    if (MBB.isReturnBlock())
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21917]: If it's not a return block, do we need to call `findInstrWhichNeedAllCSRs` and replace the terminator?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 5]
```diff
@@ -1057,17 +1236,55 @@ RISCVFrameLowering::getFrameIndexReference(const MachineFunction &MF, int FI,
...
+
+  // Functions which call __builtin_unwind_init get all their registers saved.
+  if (MF.callsUnwindInit()) {
+    SavedRegs.set();
+    return;
+  }
   if (hasFP(MF)) {
-    SavedRegs.set(RAReg);
-    SavedRegs.set(FPReg);
+    SavedRegs.set(RISCV::X1);
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1285]: What was the reason for changing this? Isn't it the same?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21878]: Doesn't callsUnwindInit need all CSR saved too?

**Review by michaelmaitland — COMMENTED** [review]
I like the idea of this patch. I left some comments as I went through it trying to understand what was going on.

File: llvm/lib/CodeGen/PrologEpilogInserter.cpp [hunk 6]
```diff
@@ -262,6 +267,7 @@ bool PEI::runOnMachineFunction(MachineFunction &MF) {
   // called functions.  Because of this, calculateCalleeSavedRegisters()
   // must be called before this function in order to set the AdjustsStack
   // and MaxCallFrameSize variables.
+  RDA->reset();
```
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/CodeGen/PrologEpilogInserter.cpp:270]: Yes, thanks.

File: llvm/lib/CodeGen/PrologEpilogInserter.cpp [hunk 7]
```diff
@@ -1164,6 +1170,7 @@ void PEI::calculateFrameObjectOffsets(MachineFunction &MF) {
 void PEI::insertPrologEpilogCode(MachineFunction &MF) {
   const TargetFrameLowering &TFI = *MF.getSubtarget().getFrameLowering();
 
+  TFI.emitCFIsForCSRsHandledByRA(MF, RDA);
```
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/CodeGen/PrologEpilogInserter.cpp:1173]: In my code, only the registers added by `RISCVFrameLowering::determineMustCalleeSaves` will be handled the same way as before (with usual prologue / epilogue).

**Comment by mgudim** [comment]
@michaelmaitland thank you for the review! I wanted this branch to be just a proof of concept, so others can see the full context for the subsequent smaller PRs, I will address your comments there.

**Comment by mgudim** [comment]
> What is the thought process of taking this approach over modifying the ShrinkWrap pass to communicate the PrologEpilogInserter where to put each callee saved register?

If you ignore the CFI stuff, this approach is much easier, simplifies things    (now we don't special treatment of CSRs in RA) and allows for more    optimizations than just shrink-wrapping by itself. Essentially, we're exposing  the code to save / restore CSRs to all the optimizations after ISel.

File: llvm/lib/CodeGen/MachineLICM.cpp [hunk 8]
```diff
@@ -262,15 +262,21 @@ namespace {
     void HoistOutOfLoop(MachineDomTreeNode *HeaderN, MachineLoop *CurLoop,
                         MachineBasicBlock *CurPreheader);
 
-    void InitRegPressure(MachineBasicBlock *BB);
+    void InitRegPressure(MachineBasicBlock *BB, const MachineLoop *Loop);
```
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/CodeGen/MachineLICM.cpp:265]: > From the [original shrink wrapping paper](https://www.cs.tufts.edu/~nr/cs257/archive/fred-chow/p85-chow.pdf):
> 
> > There is one more consideration in performing the above shrink-wrap optimization. If the shrink-wrapped region lies completely inside a loop, there will be serious performance impact: instead of saving and restoring once per invocation of the procedure, this is now repeated for each iteration of the loop. To prevent this from occurring, whenever a register is used inside a loop, we propagate its APP attribute throughout the entire region of the loop. Thus, any shrink-wrap is not allowed to penetrate loop boundaries.
> 
> I'm not sure I fully understand what's going on here but I wanted to check and see if this patch allows shrink wrapping on a per register basis inside loops. If so, is this something we should avoid? Is this concept at all related to this code?

with this patch, shrink-wrapping becomes spilling. Hopefully we never spill inside the loop.

[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/CodeGen/MachineLICM.cpp:265]: > Can you also help me understand why this patch (without this machine-licm change) leads to longer live ranges?

because on entry to the function we copy each CSR into a virtual register which will be live for the entire function.

File: llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp [hunk 9]
```diff
@@ -44,6 +44,13 @@ static cl::opt<bool> EnableLocalReassignment(
              "may be compile time intensive"),
     cl::init(false));
 
+static cl::opt<float> MinWeightRatioNeededToEvictHint(
```
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:47]: see https://github.com/llvm/llvm-project/pull/90819#discussion_r1832389172

File: llvm/lib/CodeGen/PrologEpilogInserter.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/CodeGen/PrologEpilogInserter.cpp:1173]: suppose we had the following code (before block layout is finalized):

```
ENTRY:
save x1
.cfi_offset x1, ...
...
beq ... %RET

BB:
...

RET:
ld x1, ...
.cfi_register x1, x1
...
ret
```

after the block layout the code may look like this (now `%RET` is layout successor of `%ENTRY`):


```
ENTRY:
save x1
.cfi_offset x1, ...
...
bne ... %BB

RET:
ld x1, ...
.cfi_register x1, x1

BB:
...
j %RET

...
ret
```

The cfi directives apply to all the code below it in the layout, so the cfi's are wrong now. The CFIInserter compares the CFI info from the layout successor to what it has to be and corrects it if necessary.

I basically copied that pass from the existing one `CodeGen/ CFIInstrInserter.cpp` and modified it.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21878]: > Is it possible that there are other functions in addition to these functions that need all CSRs to be saved? Is it possible that there can be multiple MI in MBB that need this kind of handling?

It's possible, I only added the ones I know.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 3 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21926]: This is a hack! I just wanted to trigger CSR saves. I hope there are better ways to do this.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 4 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21917]: We add the new fake terminator `UnreachableRET`

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 5 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1285]: I didn't mean to change this,  it's a typo.

File: llvm/lib/CodeGen/MachineLICM.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineLICM.cpp:265]: > > Can you also help me understand why this patch (without this machine-licm change) leads to longer live ranges?
> 
> because on entry to the function we copy each CSR into a virtual register which will be live for the entire function.

Was it not the case that when we spilled the CSR in PrologEpilogInserter that the register was also live for the entire function?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21878]: Do we need to be worried about not handling cases that we don't currently know about? Do we need to solve this problem more generally?

File: llvm/lib/CodeGen/MachineLICM.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/CodeGen/MachineLICM.cpp:265]: There was no single virtual register corresponding to each Callee Saved Register. The PrologEpilogEmitter approach looks after regalloc to see which callee saved registers are used in the function and spills/restores them.

File: llvm/lib/CodeGen/PrologEpilogInserter.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaelmaitland) [reply llvm/lib/CodeGen/PrologEpilogInserter.cpp:1173]: > I basically copied that pass from the existing one CodeGen/CFIInstrInserter.cpp and modified it.

Is it possible to modify `CodeGen/CFIInstrInserter.cpp` instead of having a RISC-V specific copy?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21878]: I don't know how we can solve this more generally. Maybe we can introduce a new attribute?

File: llvm/lib/CodeGen/PrologEpilogInserter.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/CodeGen/PrologEpilogInserter.cpp:1173]: Yes, it's possible. In some cases (notably when we have vector spills), we need to use cfi expression instead of simple `.cfi_offset`. I can add a small cfi-expression decoder to `CFIInserter` for that.

Here, I recorded cfi info into `RISCVMachineFunctionInfo` so it can be used by `RISCVCFIInserter`.

File: llvm/include/llvm/CodeGen/TargetFrameLowering.h [hunk 10]
```diff
@@ -210,6 +211,11 @@ class TargetFrameLowering {
   /// for noreturn nounwind functions.
   virtual bool enableCalleeSaveSkip(const MachineFunction &MF) const;
 
+  virtual void emitCFIsForCSRsHandledByRA(MachineFunction &MF,
```
Comment (qcolombet) [inline llvm/include/llvm/CodeGen/TargetFrameLowering.h:214]: We need a comment here to tell backend authors what they are supposed to do with this hook and when they should consider implementing it.

File: llvm/include/llvm/CodeGen/TargetSubtargetInfo.h [hunk 11]
```diff
@@ -317,6 +317,8 @@ class TargetSubtargetInfo : public MCSubtargetInfo {
     return false;
   }
 
+  virtual bool doCSRSavesInRA() const;
```
Comment (qcolombet) [inline llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:320]: Ditto missing comments

**Review by qcolombet — COMMENTED** [review]
I feel like there are at least three different changes in this PR and they would need to be broken down:
1. The LICM changes
2. The RDA changes
3. The CSR changes for RISC-V

I think the overall direction makes sense. I'm not super happy how the CFI are handled but I haven't given it much thoughts.

Disclaimer, I haven't looked at the RISC-V changes at all.

File: llvm/include/llvm/CodeGen/TargetFrameLowering.h (hunk 10 above)
Comment (mgudim) [reply llvm/include/llvm/CodeGen/TargetFrameLowering.h:214]: Thanks @qcolombet I'll make sure I'll add a comment when I create a separate MR for it.

**Comment by mgudim** [comment]
> I feel like there are at least three different changes in this PR and they would need to be broken down:
> 
> 1. The LICM changes
> 2. The RDA changes
> 3. The CSR changes for RISC-V
> 
> I think the overall direction makes sense. I'm not super happy how the CFI are handled but I haven't given it much thoughts.
> 
> Disclaimer, I haven't looked at the RISC-V changes at all.

Here's MR for RDA:

https://github.com/llvm/llvm-project/pull/118097

**Comment by michaelmaitland** [comment]
Any chance we could rebase this patch on main? It would enable me to run some more experiments. Currently having some trouble when I tried to rebase it myself.

**Comment by mgudim** [comment]
> Any chance we could rebase this patch on main? It would enable me to run some more experiments. Currently having some trouble when I tried to rebase it myself.

@michaelmaitland I'll do it soon, sorry for delay. I had to switch to something else for now.

**Comment by mgudim** [comment]
> > Any chance we could rebase this patch on main? It would enable me to run some more experiments. Currently having some trouble when I tried to rebase it myself.
> 
> @michaelmaitland I'll do it soon, sorry for delay. I had to switch to something else for now.

I rebased and pushed to a branch in my private repo, so we can updo the rebase if needed.

**Comment by mgudim** [comment]
> > > Any chance we could rebase this patch on main? It would enable me to run some more experiments. Currently having some trouble when I tried to rebase it myself.
> > 
> > 
> > @michaelmaitland I'll do it soon, sorry for delay. I had to switch to something else for now.
> 
> I rebased and pushed to a branch in my fork (https://github.com/mgudim/llvm-project/tree/rebased_save_csr_in_ra), so we can updo the rebase if needed.

In my original patch there was a hack related to emission of CFI's and the CFI inserter. I am trying to get rid of that we can get rid of the RISCV CFI inserter.

**Comment by michaelmaitland** [comment]
> In my original patch there was a hack related to emission of CFI's and the CFI inserter. I am trying to get rid of that we can get rid of the RISCV CFI inserter.

I **strongly** support this direction. Copying the target agnostic CFI inserter as a target specific CFI inserter is likely not a viable direction. If we need to add some target specific logic to the CFI inserter, the we should use target hooks.

**Comment by mgudim** [comment]
> > In my original patch there was a hack related to emission of CFI's and the CFI inserter. I am trying to get rid of that we can get rid of the RISCV CFI inserter.
> 
> I **strongly** support this direction. Copying the target agnostic CFI inserter as a target specific CFI inserter is likely not a viable direction. If we need to add some target specific logic to the CFI inserter, the we should use target hooks.

I make the patch use the main CFIInserter and got rid of the RISCVCFIInserter completely. I also ran spec for scalar and everything passed. Running with vector now.

**Comment by asb** [comment]
@mgudim @ppenzin I was interested in kicking the tires on this to see impact on recentish HEAD on SPEC. Do either of you have a recent version of this patchset that is known good? I built https://github.com/mgudim/llvm-project/tree/save_csr_in_ra3 but I'm getting compiler crashes at the initial cmake configure for llvm-test-suite with `-mllvm -riscv-save-csrs-early=true`, or when I do it with `-save-temps=obj` as well I get errors about unrecognized instruction mnemonic `BUNDLE`.

**Comment by mgudim** [comment]
> @mgudim @ppenzin I was interested in kicking the tires on this to see impact on recentish HEAD on SPEC. Do either of you have a recent version of this patchset that is known good? I built https://github.com/mgudim/llvm-project/tree/save_csr_in_ra3 but I'm getting compiler crashes at the initial cmake configure for llvm-test-suite with `-mllvm -riscv-save-csrs-early=true`, or when I do it with `-save-temps=obj` as well I get errors about unrecognized instruction mnemonic `BUNDLE`.

@asb I think this branch:
https://github.com/mgudim/llvm-project/tree/rebased_save_csr_in_ra
should be working. To enable saving csrs early don't forget to add these two flags:

```
-mllvm -riscv-enable-save-csr-in-ra=true 
-mllvm -min-weight-ratio-needed-to-evict-hint=7.5
```

That branch only has one huge commit and its' just POC. I am still working on breaking it up into smaller commits and putting them on (https://github.com/mgudim/llvm-project/tree/save_csr_in_ra3). But that branch is in broken state right now

**Comment by asb** [comment]
> > @mgudim @ppenzin I was interested in kicking the tires on this to see impact on recentish HEAD on SPEC. Do either of you have a recent version of this patchset that is known good? I built https://github.com/mgudim/llvm-project/tree/save_csr_in_ra3 but I'm getting compiler crashes at the initial cmake configure for llvm-test-suite with `-mllvm -riscv-save-csrs-early=true`, or when I do it with `-save-temps=obj` as well I get errors about unrecognized instruction mnemonic `BUNDLE`.
> 
> @asb I think this branch: https://github.com/mgudim/llvm-project/tree/rebased_save_csr_in_ra should be working. To enable saving csrs early don't forget to add these two flags:
> 
> ```
> -mllvm -riscv-enable-save-csr-in-ra=true 
> -mllvm -min-weight-ratio-needed-to-evict-hint=7.5
> ```
> 
> That branch only has one huge commit and its' just POC. I am still working on breaking it up into smaller commits and putting them on (https://github.com/mgudim/llvm-project/tree/save_csr_in_ra3). But that branch is in broken state right now

Thank you. I haven't looked to see what it would take to rebase on current HEAD, but comparing that branch with a baseline of commit `d7e48fbf205a01fcbc109b2555b12aa0d37845a4` (the last upstream commit before your downstream patches), and running SPEC -march=rva22u64 I get these results for dynamically executed instructions:

```
Benchmark                       Baseline            csrinra   Diff (%)
============================================================
500.perlbench_r         180245021535    174761106584     -3.04%
502.gcc_r               220872098438    204171725268     -7.56%
505.mcf_r               131589945206    132303362396      0.54%
508.namd_r              220648060756    219744519650     -0.41%
510.parest_r            291341819989    291293000597     -0.02%
511.povray_r             31914276340     27581024333    -13.58%
519.lbm_r                94166321455     96148444880      2.10%
520.omnetpp_r           138002605450    132999926345     -3.63%
523.xalancbmk_r         283566181775    269486838918     -4.97%
525.x264_r              380165034541    383766253106      0.95%
526.blender_r           660528269650    645308870803     -2.30%
531.deepsjeng_r         355058534696    357485665398      0.68%
538.imagick_r           238573640980    238240294103     -0.14%
541.leela_r             421886351049    411685513225     -2.42%
544.nab_r               415595727842    415394887715     -0.05%
557.xz_r                132548717793    133865723492      0.99%
```

There's some big wins here, especially for povray and gcc.

Based on comments in this thread there may be issues with the wider llvm-test-suite, but all SPEC benchmarks compile and run to completion. If you're planning to do a new rebase do let me know as that would be a better test to run.

**Comment by asb** [comment]
> As previously requested, I've taken the contents of @mgudim's branch and made a PR stack:
> 
>     * PR 1/5: [[WIP][CodeGen][RISCV] Allow multiple location for the same CSR. #170606](https://github.com/llvm/llvm-project/pull/170606)
> 
>     * PR 2/5: [[WIP][CodeGen][DebugInfo][RISCV] Support scalable offsets in CFI #170607](https://github.com/llvm/llvm-project/pull/170607)
> 
>     * PR 3/5: [[WIP][CodeGen] Split `determineCalleeSaves` #170609](https://github.com/llvm/llvm-project/pull/170609)
> 
>     * PR 4/5: [[WIP][CodeGen] Enable early CFI #170610](https://github.com/llvm/llvm-project/pull/170610)
> 
>     * PR 5/5: [[WIP][CodeGen] Allow register allocator to drive save/restore #170611](https://github.com/llvm/llvm-project/pull/170611)
> 
> 
> I've omitted one "fix a test' commit and a couple of commits are swapped by accident. It is also rebased on a more recent `main`. PR 2 has clean CI and can be probably extracted out, though might require additional DWARF verification.

Thank you! The changes to ‎llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp‎ to add support for `-min-weight-ratio-needed-to-evict-hint` seem to be gone from this new patch stack. Was that intentional?

**Comment by ppenzin** [comment]
> That branch only has one huge commit and its' just POC. I am still working on breaking it up into smaller commits and putting them on (https://github.com/mgudim/llvm-project/tree/save_csr_in_ra3). But that branch is in broken state right now

That would explain the situation with the PR stack. I thought the branch was ready to upstream, didn't check the comment before working on it.

**Comment by bojle** [comment]
Hi @mgudim,

Consider this simple example that contains just the frame record for prologue: https://godbolt.org/z/7o1jEe9s5

Both clang (trunk) and gcc (trunk) (as seen in compiler explorer) sink the prologue and keep the hot path clean. 

When I run this example with this patch, I am seeing that frame record is being nailed at entry:

```
; -target riscv64 -O3 -fno-pic -mllvm --riscv-enable-save-csr-in-ra
f3():
        addi    sp, sp, -16 ; this is sunk in trunk
        lui     a0, %hi(cond)
        ld      a0, %lo(cond)(a0)
        sd      s1, 8(sp)
        mv      s1, ra
        beqz    a0, .LBB2_2
        call    ext()
.LBB2_2:
        li      a0, 42
        mv      ra, s1
        ld      s1, 8(sp)
        addi    sp, sp, 16
        ret
```
Similar behavior on aarch64 (I've ported this patch locally). This could certainly lead to regressions.
